### PR TITLE
remove object class from base classes in type calls

### PR DESCRIPTION
**Pull request by lev-blit**
This solves #532

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -707,3 +707,10 @@ Availability:
 -def f(x: 'queue.Queue[int]') -> C:
 +def f(x: queue.Queue[int]) -> C:
 ```
+
+
+### remove inheriting from object when creating classes with `type()`
+```diff
+-A = type("A", (object,), {})
++A = type("A", (), {})
+```
```
Comment (asottile) [inline README.md:716]: can you place this next to the other relevant `object` thing?

File: tests/features/type_bases_object_test.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        ('A = type("A", (object,), {})', 'A = type("A", (), {})'),
+        ('B = type("B", (object, tuple), {})', 'B = type("B", (tuple,), {})'),
+        (
+            'C = type("C", (object, foo, bar), {})',
+            'C = type("C", (foo, bar), {})',
+        ),
+        ('D = type("D", (tuple, object), {})', 'D = type("D", (tuple,), {})'),
+        (
+            'E = type("E", (foo, bar, object), {})',
+            'E = type("E", (foo, bar), {})',
```
Comment (asottile) [inline tests/features/type_bases_object_test.py:33]: I had initially attempted this rewrite as well but ran into all sorts of edge cases where the tuple spans more than one line and/or has trailing commas -- can you include some tests for that case as well

File: pyupgrade/_token_helpers.py [hunk 3]
```diff
@@ -352,6 +352,60 @@ def remove_base_class(i: int, tokens: list[Token]) -> None:
         del tokens[left:last_part + 1]
 
 
+def remove_base_class_from_type_call(i: int, tokens: list[Token]) -> None:
+    # look forward and backward to find commas / parens
```
[the remark this answers is not in the document]
Comment (lev-blit) [reply pyupgrade/_token_helpers.py:356]: Understood, added arguments to `remove_base_class` to support the difference between the cases

File: pyupgrade/_plugins/type_bases_object.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        should_move_right_edge=_should_move_right_edge,
+        not_right_edge=_not_right_edge,
+        last_part_function=_last_part_function,
+        do_brace_stack_pop_loop=False,
+        remove_bases=_remove_bases,
+        single_base_char=',',
+        multiple_first_char=')',
+        multiple_first=_multiple_first,
+        multiple_last=_multiple_last,
+    )
```
Comment (asottile) [inline pyupgrade/_plugins/type_bases_object.py:85]: this is a maintenance nightmare, I can't accept the code in this state

Comment (lev-blit) [reply pyupgrade/_plugins/type_bases_object.py:85]: well there is a significant difference between `A = type(...)` and `class A(...)` so even though most of the logic is the same, you can see for yourself how much differs(or at least how much differs in my implementation)
this is either something that can be solved by entirely separating the cases(which will cause code duplication) or by someone more experienced / more adapt at this sort of thing tackling the issue.

do you think this is something that I should keep working on, even though it may take a few more tries(if I even find a better solution with my current skill set), or should we close the PR here and leave this for someone more experienced to tackle?

Comment (asottile) [reply pyupgrade/_plugins/type_bases_object.py:85]: > well there is a significant difference between A = type(...) and class A(...)

then that should be a dead giveaway that they should **NOT** share code

Comment (lev-blit) [reply pyupgrade/_plugins/type_bases_object.py:85]: well that is what I thought at first but you requested that they do(unless I misunderstood your comment as to which function that is)

"this function should reuse the function call parsing function rather than duplicating its work"

Comment (lev-blit) [reply pyupgrade/_plugins/type_bases_object.py:85]: well I agreed with that, but it seems I have misunderstood a comment of yours, understanding that you wanted them to share code(said comment - "this function should reuse the function call parsing function rather than duplicating its work")
😞

Comment (lev-blit) [reply pyupgrade/_plugins/type_bases_object.py:85]: thank you and sorry for the misunderstanding, will use it :)

**Comment by lev-blit** [comment]
@asottile what should be done for `(foo, object, bar)` and similar cases where object isn't first or last? This doesn't seem like something anyone should write, but maybe it should be supported(or on the contrary, explicitly unsupported)

**Comment by asottile** [comment]
so the weird thing is `type('a', (), {})` in python 2.x also makes a new-style class so it also seems like something someone shouldn't write too:

```pycon
>>> x = type('a', (), {})
>>> x.__bases__
(<type 'object'>,)
```

I would probably remove it anyway 🤷

File: pyupgrade/_plugins/type_bases_object.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+from pyupgrade._data import TokenFunc
+from pyupgrade._token_helpers import find_open_paren
+from pyupgrade._token_helpers import parse_call_args
+
+
+def remove_base_class_from_type_call(i: int, tokens: list[Token]) -> None:
+    type_start = find_open_paren(tokens, 0)
+    bases_start = find_open_paren(tokens, type_start + 1)
+    bases, end = parse_call_args(tokens, bases_start)
+    for base_start, base_end in bases:  # pragma: no cover
```
Comment (asottile) [inline pyupgrade/_plugins/type_bases_object.py:22]: can't ship with this -- means you haven't tested everything

this function is way too complicated -- I don't think it's correct either.  you should also be able to use the original `object` base index without needing to re-find it here and utilize the argument removal functions

File: pyupgrade/_plugins/type_bases_object.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+from pyupgrade._data import register
+from pyupgrade._data import State
+from pyupgrade._data import TokenFunc
+from pyupgrade._token_helpers import find_open_paren
+from pyupgrade._token_helpers import parse_call_args
+
+
+def remove_base_class_from_type_call(i: int, tokens: list[Token]) -> None:
+    type_start = find_open_paren(tokens, 0)
+    bases_start = find_open_paren(tokens, type_start + 1)
```
Comment (asottile) [inline pyupgrade/_plugins/type_bases_object.py:20]: this isn't necessarily correct -- `type(class_name(), (...))`

File: pyupgrade/_plugins/type_bases_object.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                    if len(bases) == 2:
+                        # remove the trailing comma,
+                        # there is one after the first base
+                        if tokens[base_end].src == ',':
+                            del tokens[base_start:base_end + 1]
+                        else:
+                            del tokens[base_start:base_end]
+                    # (foo, bar, ..., object[,]) -> (foo, bar, ...[,])
+                    else:
+                        if tokens[base_end - 1].name == UNIMPORTANT_WS and \
```
Comment (asottile) [inline pyupgrade/_plugins/type_bases_object.py:57]: don't use backslashes

**Comment by lev-blit** [comment]
I won't have much time to work on this anymore :/ -- closing the PR so others might take a stab at this

p.s. thank you for being patient with me through this (my first contribution), this was an educating experience :)
